File: solutions/Leetcode_658/leetcode_658.cc
#include <cmath>
#include <deque>
#include <vector>

using namespace std;


class Solution1 {
public:
    vector<int> findClosestElements(vector<int>& arr, int k, int x) {
        int left = 0, right = arr.size() - k;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (x - arr[mid] > arr[mid+k] - x)
                left = mid + 1;
            else
                right = mid;
        }
        return vector<int>(arr.begin() + left, arr.begin() + left + k);
    }
};


class Solution2 {
public:
    vector<int> findClosestElements(vector<int>& arr, int k, int x) {
        deque<int> ret;
        for (int& y : arr) {
            if (ret.size() < k)
                ret.push_back(y);
            else if (abs(x - y) < abs(x - ret.front())) {
                ret.pop_front();
                ret.push_back(y);
            }
        }
        return vector<int>(ret.begin(), ret.end());
    }
};

File: solutions/Leetcode_658/leetcode_658.py
from typing import List


class Solution1:
    def find_closest_elements(self, arr, k, x):
        left = 0
        right = len(arr) - k
        while left < right:
            mid = left + (right - left) // 2
            if x - arr[mid] > arr[mid+k] - x:
                left = mid + 1
            else:
                right = mid
        return arr[left:left+k]


class Solution2:
    def find_closest_elements(self, arr: List[int], k: int, x: int) -> List[int]:
        closest_elements = []
        for y in arr:
            if len(closest_elements) < k:
                closest_elements.append(y)
            elif abs(y - x) < abs(closest_elements[0] - x):
                closest_elements.remove(closest_elements[0])
                closest_elements.append(y)
        return closest_elements
